src/rust/engine/rule_graph/src/rules.rs

```diff
@@ -22,26 +23,61 @@ pub trait TypeId:
     I: Iterator<Item = Self>;
 }
 
-pub trait DependencyKey:
-  Clone + Copy + Debug + Display + Hash + Ord + Eq + Sized + 'static
-{
-  type TypeId: TypeId;
+#[derive(DeepSizeOf, Eq, Hash, PartialEq, Clone, Debug, PartialOrd, Ord)]
+pub struct DependencyKey<T: TypeId> {
+  product: T,
+  provided_params: SmallVec<[T; 2]>,
```

Comment:
I'm not following why it's set to 2?

Reply:
Smallvec has fixed "on stack" storage that it uses for small element counts. In this case, <= 2 elements will not allocate, and instead will be stored inline in the struct. I chose 2 in this case just because the cases we're talking about using "multiple parameters to `Get`" with all involve 2 parameters.

Reply:
Probably worth a comment de-magicifying that